### ISSUE TITLE
Training stop at about every 200 epochs~

### ISSUE THREAD

**Issue by YJHMITWEB** [tensorpack/tensorpack#1111]
If you're asking about an unexpected problem which you do not know the root cause,
use this template. __PLEASE DO NOT DELETE THIS TEMPLATE, FILL IT__:

If you already know the root cause to your problem,
feel free to delete everything in this template.

### 1. What you did:

(1) **CUDA_VISIBLE_DEVICES=0,1,2,3 python ./train.py --config MODE_MASK=True MODE_FPN=True DATA.BASEDIR=/home/data/COCO BACKBONE.WEIGHTS=./ckpt/ImageNet-R50-AlignPadding.npz**
I'm running example->FasterRCNN train.

(2) **If you're using examples, have you made any changes to the examples? Paste `git diff` here:**
modify config.py:
 `_C.TRAIN.NUM_GPUS = None` to `_C.TRAIN.NUM_GPUS = 4`, since I'm using 4 Titan V.

(3) **If not using examples, tell us what you did:**

  It's always better to copy-paste what you did than to describe them.

  Please try to provide enough information to let other __reproduce__ your issues. 
  Without reproducing the issue, we may not be able to investigate it.

### 2. What you observed:

(1) **[0317 00:44:32 @monitor.py:467] rpn_losses/level4/box_loss: 0.0053975
[0317 00:44:32 @monitor.py:467] rpn_losses/level4/label_loss: 0.0060109
[0317 00:44:32 @monitor.py:467] rpn_losses/level4/label_metrics/precision_th0.1: 0.56266
[0317 00:44:32 @monitor.py:467] rpn_losses/level4/label_metrics/precision_th0.2: 0.61337
[0317 00:44:32 @monitor.py:467] rpn_losses/level4/label_metrics/precision_th0.5: 0.76888
[0317 00:44:32 @monitor.py:467] rpn_losses/level4/label_metrics/recall_th0.1: 0.78846
[0317 00:44:32 @monitor.py:467] rpn_losses/level4/label_metrics/recall_th0.2: 0.77893
[0317 00:44:32 @monitor.py:467] rpn_losses/level4/label_metrics/recall_th0.5: 0.75887
[0317 00:44:32 @monitor.py:467] rpn_losses/level4/num_pos_anchor: 5.0245
[0317 00:44:32 @monitor.py:467] rpn_losses/level4/num_valid_anchor: 12.535
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/box_loss: 0.0047516
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/label_loss: 0.0033548
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/label_metrics/precision_th0.1: 0.73172
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/label_metrics/precision_th0.2: 0.74464
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/label_metrics/precision_th0.5: 0.79031
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/label_metrics/recall_th0.1: 0.84738
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/label_metrics/recall_th0.2: 0.84368
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/label_metrics/recall_th0.5: 0.84144
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/num_pos_anchor: 6.2395
[0317 00:44:32 @monitor.py:467] rpn_losses/level5/num_valid_anchor: 7.4686
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/box_loss: 0.0048501
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/label_loss: 0.00026581
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/label_metrics/precision_th0.1: 0.7694
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/label_metrics/precision_th0.2: 0.76942
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/label_metrics/precision_th0.5: 0.76942
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/label_metrics/recall_th0.1: 0.76948
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/label_metrics/recall_th0.2: 0.76948
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/label_metrics/recall_th0.5: 0.76948
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/num_pos_anchor: 3.1098
[0317 00:44:32 @monitor.py:467] rpn_losses/level6/num_valid_anchor: 3.172
[0317 00:44:32 @monitor.py:467] sample_fast_rcnn_targets/num_bg: 443.58
[0317 00:44:32 @monitor.py:467] sample_fast_rcnn_targets/num_fg: 68.418
[0317 00:44:32 @monitor.py:467] sample_fast_rcnn_targets/proposal_metrics/best_iou_per_gt: 0.76231
[0317 00:44:32 @monitor.py:467] sample_fast_rcnn_targets/proposal_metrics/recall_iou0.3: 0.98801
[0317 00:44:32 @monitor.py:467] sample_fast_rcnn_targets/proposal_metrics/recall_iou0.5: 0.97087
[0317 00:44:32 @monitor.py:467] total_cost: 0.9818
[0317 00:44:32 @monitor.py:467] wd_cost: 0.30621
[0317 00:44:32 @base.py:274] Start Epoch 220 ...
 48%|######################################8                                          |240/500[02:00<01:53, 2.28it/s]**

Starting from 0 epoch, it works perfectly fine until about 220 epochs (I test it several times, 220 epochs is the average), it just stacks there without raising any errors.
Using `nvidia-smi` command, all 4 Titan V are in idle, but their memory is all taken over, about 9000 MB of a total 12000 MB per GPU.

It's always better to copy-paste what you observed instead of describing them.

It's always better to paste **as much as possible**, although sometimes a partial log is OK.

Tensorpack typically saves stdout to its training log.
If stderr is relevant, you can run a command with `CMD 2>&1 | tee logs.txt`
to save both stdout and stderr to one file.

(2) **Other observations, if any:**
GPU utilization turns into 0% at about 220 epochs.

### 3. What you expected, if not obvious.

If you expect higher speed, please read
http://tensorpack.readthedocs.io/tutorial/performance-tuning.html 
before posting.

If you expect certain accuracy, only in one of the two conditions can we help with it:
(1) You're unable to match the accuracy documented in tensorpack examples.
(2) It appears to be a tensorpack bug.

Otherwise, how to train a model to certain accuracy is a machine learning question and is
not our responsibility to figure out.

### 4. Your environment:
  + Python version: 3.6
  + TF version: 1.10
  + Tensorpack version: 0.9.1
  + Ubuntu 16.04.
  + CUDA 9.0, CUDNN 7.0
  + 4 Titan V with 12 GB memory each.
  + NCCL installed, python-prctl installed

The training process works fine by observing the evaluation results every 25 epochs, and when it stacks at about 220 epochs, I can simply stop and kill the process, then restart it with loading the last checkpoint. I just cannot figure out what may cause the stack.

**Comment by ppwwyyxx** [comment]
That is weird. Did you make other changes (about trainer or data)?
I have not seen this with the default trainer and don't have an idea why this may happen.

**Comment by YJHMITWEB** [comment]
I did not make any other changes to any files under the `example` path.
It's weird.... Just that I have used tensorflow to develop for like more than 3 years, but haven't meet this kind of thing. 
Maybe there are some implicit problems in my machine, really have no idea...
But right now, I could just restart it every 220 epochs, it won't make any trouble.

**Comment by YJHMITWEB** [comment]
By the way, Tensorpack is a really great work (you paper GroupNorm also)!!!!（膜拜）

**Comment by ppwwyyxx** [comment]
It is worth trying to upgrade tensorflow. I've seen some weird errors (unlike yours) in the past but haven't seen any recently.

**Comment by YJHMITWEB** [comment]
Sure, I just upgrade tensorflow to 1.12, and now testing whether the problem still exists or gone.

**Comment by YJHMITWEB** [comment]
Unfortunately, the problem still exists..
Every time after it runs abut 220 epochs, the whole process stops.
I start from scratch, 0 epoch, and it stops at 220 epochs,
then restart it, load the last checkpoint, start from 220, then it stops again at 439 epochs,
then 648 epochs...
Also, I have found that as the training processes, the memory usage gradually increases (about 0.3~0.4% per epoch of a total 256 GB memory). I'm not sure if it's because some dynamic nodes defined in the graph, and too many of them are generated after a long-time graph running, which may consume too much memory, leading to the crash...

**Comment by ppwwyyxx** [comment]
Then your issue sounds similar to #1097 : memory gradually increases, and lead to a DeadlineExceedErrror (which is somewhat similar to getting stuck). However there, the issue does not appear for COCO dataset.

One possibility is to replace `MultiProcessMapData` in `data.py` by `MultiThreadMapData` or `MapData` (probably will be slow) -- this may make a difference, though I'm still not sure what the issue is at the moment.

**Comment by YJHMITWEB** [comment]
Ok, I will modify the `data.py` and test again.

**Comment by ppwwyyxx** [comment]
Another thing to try is to run `python data.py`. It will just iterate over the the entire dataflow for 50000 iterations (which you can change to more iterations). You can use this to test whether the dataflow will grow the memory usage indefinitely.

**Comment by YJHMITWEB** [comment]
.........I've tried a lot, replacing `MultiProcessMapData` in `data.py` by `MultiThreadMapData`, using less threads in reading data, the memory consumption is now normal, but still, every thing stops at every 220 epochs...

**Comment by Xiao-OMG** [comment]
@YJHMITWEB  I train a classification network on my own dataset and  have the same problems: 
 1. Memory gradually increases,
 2. Training stopped without any error raising. It stopped at the final step of 10th evaluation.  
![image](https://user-images.githubusercontent.com/18168264/54814792-63ec1880-4ccb-11e9-8715-4e4c0b09e663.png)

I will try MultiThreadMapData. Hope it help to solve problem 1

My TF is 1.9.0, maybe i should try upgrade TF

**Comment by Xiao-OMG** [comment]
Emmm...
I used  **MultiThreadMapData** to generate train_data_flow, but **MultiProcessMapData** for eval_data_flow.  It stopped at the first step of the first evaluation........
Maybe it should use the same function

**Comment by ppwwyyxx** [comment]
@Xiao-OMG Does the issue happen if you remove your evaluation?
I thinks it's actually getting stuck at the last step of the training, not at the evaluation.
The last step of an epoch often includes some extra work to do, e.g., running the summary ops and writing the summaries, which makes it different from the other steps.
If you can create and share a reproducible example it would be great.

**Comment by ppwwyyxx** [comment]
Though we still can't reproduce or understand why it stucks after training for a while, there have been several recent issues about memory use in the Mask R-CNN example, which leads to weird issues. The memory issue mainly came from the use of `fork` in a multi threaded process, which is by itself unsafe and may lead to weird issues.
After https://github.com/tensorpack/tensorpack/commit/8f10b0f8580293721b07a40e3e9d4c6d82398790, the Mask R-CNN example now switch to "spawn" multiprocessing method which is safer and reduce CPU memory usage by about 2x. Hopefully this can resolve some of the issues.

Closing due to lack of activity, but feel free to comment if there is any follow ups about the issue.

**Comment by ppwwyyxx** [comment]
I saw a similar incident recently that in one old environment, all training jobs get stuck at a certain epoch. But it was resolved after I upgrade cuda from 9.0 to 9.2 (and install the corresponding tensorflow). This may be related.

**Comment by Xiao-OMG** [comment]
thinks, now i use `MultiThreadMapData` , it has not happened yet.
